x/config/builder.go

```diff
@@ -0,0 +1,265 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package config
+
+import (
+	"fmt"
+
+	"go.uber.org/yarpc"
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/internal/errors"
+)
+
+// buildTransport builds a Transport from the given value. This will panic if
+// the output type is not a Transport.
+func buildTransport(cv *configuredValue) (transport.Transport, error) {
+	result, err := cv.Build()
+	if err != nil {
+		return nil, err
+	}
+	return result.(transport.Transport), nil
+}
+
+// buildInbound builds an Inbound from the given value. This will panic if the
+// output type for this is not transport.Inbound.
+func buildInbound(cv *configuredValue, t transport.Transport) (transport.Inbound, error) {
+	result, err := cv.Build(t)
+	if err != nil {
+		return nil, err
+	}
+	return result.(transport.Inbound), nil
+}
+
+// buildUnaryOutbound builds an UnaryOutbound from the given value. This will panic
+// if the output type for this is not transport.UnaryOutbound.
+func buildUnaryOutbound(cv *configuredValue, t transport.Transport) (transport.UnaryOutbound, error) {
+	result, err := cv.Build(t)
+	if err != nil {
+		return nil, err
+	}
+	return result.(transport.UnaryOutbound), nil
+}
+
+// buildOnewayOutbound builds an OnewayOutbound from the given value. This will
+// panic if the output type for this is not transport.OnewayOutbound.
+func buildOnewayOutbound(cv *configuredValue, t transport.Transport) (transport.OnewayOutbound, error) {
+	result, err := cv.Build(t)
+	if err != nil {
+		return nil, err
+	}
+	return result.(transport.OnewayOutbound), nil
+}
+
+type configuredClient struct {
+	Service string
+	Unary   *configuredOutbound
+	Oneway  *configuredOutbound
+}
+
+type configuredInbound struct {
+	Transport string
+	Value     *configuredValue
+}
+
+type configuredOutbound struct {
+	Transport string
+	Value     *configuredValue
+}
+
+type builder struct {
+	Name string
+
+	// Transports that we actually need and their specs. We need a transport
+	// only if we have at least one inbound or outbound using it.
+	needTransports map[string]*compiledTransportSpec
+
+	transports map[string]*configuredValue
+	inbounds   []configuredInbound
+	clients    map[string]*configuredClient
+}
+
+func newBuilder(name string) *builder {
+	return &builder{
+		Name:           name,
+		needTransports: make(map[string]*compiledTransportSpec),
+		transports:     make(map[string]*configuredValue),
+		clients:        make(map[string]*configuredClient),
+	}
+}
+
+func (b *builder) Build() (yarpc.Config, error) {
+	transports := make(map[string]transport.Transport)
+
+	for name, spec := range b.needTransports {
+		cv, ok := b.transports[name]
+
+		var err error
+		if !ok {
+			// No configuration provided for the transport. Use an empty map.
+			cv, err = spec.Transport.Decode(attributeMap{})
+			if err != nil {
+				return yarpc.Config{}, err
+			}
+		}
+
+		transports[name], err = buildTransport(cv)
```

Comment:
Not at this time. You *could* register a TransportSpec where those things are already injected somehow (`cfg.RegisterTransport(mytransport.TransportSpec(logger))`) but it's not the best experience.

Being able to inject arbitrary dependencies to arbitrary subcomponents sounds too much like a dependency injection container—a problem I would rather not solve here. Maybe once `dig` is moved to its own repo, this component can rely on it rather than the hard coded dependency graph where the inbounds and outbounds depend on the transport.